### DISCUSSION TITLE
How to live graph in python [mjvFigure]?

### DISCUSSION THREAD

**Discussion by rohit-kumar-j** [google-deepmind/mujoco#298] [Asking for Help] [answered]
I am trying to build a graphing tool built on top of [mujoco-python-viewer](https://github.com/rohanpsingh/mujoco-python-viewer) for [pid tuning](https://vimeo.com/246182184) ,etc. Intending to use it like this:
```python
model = mujoco.MjModel.from_xml_path("path_to_file")
data = mujoco.MjData(model)
while True:
    viewer.graph_1(self.data.qpos[0])
    viewer.graph_1(actuation_signal) # actuation_signal is a data stream of float values (ex: from a loop)
    viewer.render()
```
I have looked at [`simulate.cc`](https://github.com/deepmind/mujoco/blob/main/sample/simulate.cc), but was not able to obtain the required result using [profiler_update section](https://github.com/deepmind/mujoco/blob/0e5d062302de3527c6dee766af4efcae78e5f532/sample/simulate.cc#L423).

1. How do we populate and call the graphing utilities in python for the live graph of joint values similar to the CPU-time profiler:

![image](https://user-images.githubusercontent.com/37873142/170141477-4850cb0e-ce68-4712-9c2c-93fa08ebe8c2.png)

How to write to these pixels?
![image](https://user-images.githubusercontent.com/37873142/170191110-5016b7c9-7979-45cd-af99-adb4ebe43383.png)

**Comment by JeanElsner** [comment]
It's probably easier and more pythonic to simply use a Python library to do the plotting. For example we use animated `matplotlib` widgets within a Qt UI to do live plots, adjust parameters etc.

**Comment by nimrod-gileadi** [comment, marked as answer]
I agree with @JeanElsner's answer, but if you do want to use mjvFigure, it should be possible.
Please share the code you used to try to replicate `profiler_update`.

I'd follow what happens in simulate.cc to `figtimer`, and check that your code does all of it:

1. It's initialized with mjv_defaultFigure - this will be done for you automatically by the Python bindings.
2. Various properties are set on it for axes, colours, etc.
3. `figtimer.linepnt` and `figtimer.linedata` are set.
4. On every frame, `profilershow` is called, which calls `mjr_figure` with a particular viewport and the render context. This would be `mujoco.mjr_figure` in python.

**Reply by rohit-kumar-j** [reply]
What exactly are `linepnt` and `linedata`? I went through the documentation but could not understand how it is actually used.

**Reply by yuvaltassa** [reply]
You are absolutely right. While MuJoCo has great documentation for most things, plotting (and UI) rendering is an exception. We intend to fix this very soon with comprehensive documentation, but in the meantime [here](https://github.com/deepmind/mujoco/blob/main/include/mujoco/mjvisualize.h#L344) is the type definition and [here](https://github.com/deepmind/mujoco/blob/main/src/render/render_gl2.c#L737) is the 2D plot implementation. I apologise that this is the best we can do for now.

**Reply by rohit-kumar-j** [reply]
@yuvaltassa What is `float   linedata[mjMAXLINE][2*mjMAXLINEPNT]; // line data (x,y)`? Is it [x][y] values as in:
```python
pixel_value = figure.linedata[x][y] # either 1 or 0
```
With a resolution of 2000*100 `(Y*X)` in pixels per graph?

**Reply by yuvaltassa** [reply]
You can draw up to `mjMAXLINE` lines, each line with at most `mjMAXLINEPNT` points. The XY values values of the data are in `linedata`. The memory space is preallocated with the maximum values. The actual lines that you want to plot is given by
```C
int     linepnt[mjMAXLINE];     // number of points in line; (0) disable
```
Each integer in `linepnt[i]` is either 0 (don't draw line `i`) or `n>0` (draw `n` data points from line `i`).
Makes sense?

**Reply by rohit-kumar-j** [reply]
Thank you, I was able to figure that out now, I have been working with the following lines. The `[i]`,`[i+1]`,`[i-1]`,`[i+2]`,`[i-2]` is not clear though. 

How do we write to the `[2*mjMAXLINEPNT]`?

```python
self.fig.linedata[0][i] = sensor_data[0] #sensor 0 data
self.fig.linedata[0][i - 1] = 0.0
self.fig.linedata[0][i + 1] = 0.0
self.fig.linedata[0][i - 2] = 0.0
self.fig.linedata[0][i + 2] = 0.0
```

**Reply by rohit-kumar-j** [reply]
Is this layout correct?:
With O as (0,0), and point B addressed something like:
```python
# B is at(2,-1) relative to (0,0)
b_xpos = 2 
b_ypos = -1

# write data
fig.linedata[line_id][4] = sensor_data[line_id]
fig.linedata[line_id][-1] = sensor_data[line_id]

# clear previous data (unclear)
fig.linedata[line_id][b_ypos-2] = 0.0
fig.linedata[line_id][b_ypos-1] = 0.0

```
![square](https://user-images.githubusercontent.com/37873142/170822909-e16593d2-c793-45aa-be87-636680b5ff4d.png)

**Comment by rohit-kumar-j** [comment]
Thanks for your answers!
@JeanElsner I did try [blitting ](https://alexgude.com/blog/matplotlib-blitting-supernova/)with `matplotlib`, but wanted to add more functionality to interact with the simulation. like an autotuning algorithm for PID, so I wanted to use the inbuilt plotter in mujoco.

@nimrod-gileadi This is the code for the graph I have used(I can provide the entire code, but most of it might not be necessary I think):
```python
    def __init__(self, model, data):
        self.fig = mujoco.MjvFigure()
        # Manual
        self.fig.range[0][0] = -5
        self.fig.range[0][1] = 5
        self.fig.range[1][0] = -5
        self.fig.range[1][1] = 5
        self.fig.flg_extend = 0
        self.fig.flg_symmetric = 0
        self.pid_viewport = mujoco.MjrRect(0, 0, 800, 200)


    def sensorupdate(self):
        num_sensors = self.model.nsensor
        sensor_data = self.data.qpos
        adr = self.model.sensor_adr
        print("num_sensors: ", num_sensors, sensor_data, adr)
        p = self.fig.linepnt[0]
        pnt = mujoco.mju_min(201, self.fig.linepnt[0] + 1)
        # print(f"pnt: {pnt}")
        for n in range(0, 3):
            # shift data
            for i in range(1, 0):
                self.fig.linedata[n][2 * i + 1] = self.fig.linedata[n][2 * i - 1]
                print(f"n:{n} i:{i}")

            # assign new
            self.fig.linepnt[n] = pnt
            self.fig.linedata[n][1] = sensor_data[n]
        print(self.fig.linedata)

    # show sensor figure
    def sensorshow(self, rect):
        mujoco.mjr_figure(self.pid_viewport, self.fig, self.callbacks.ctx)

    def render(self): #This is called in each loop
        self.sensorupdate()
        self.sensorshow(rect=self.pid_viewport)
        mujoco.mjr_overlay(
                        mujoco.mjtFontScale.mjFONTSCALE_100,
                        mujoco.mjtGridPos.mjGRID_BOTTOMLEFT,
                        self.pid_viewport,
                        "t1",
                        "t2",
                        self.callbacks.ctx,
                    )
```

https://user-images.githubusercontent.com/37873142/170640093-95a3c8c3-9183-47b2-b435-25c70a1285e9.mp4

**Comment by rohit-kumar-j** [comment]
@nimrod-gileadi @yuvaltassa Thanks for the support! I was able to figure out how it works. 
* I needed to set some of the axes to automatic scaling
* The math according to the [`simulate.cc #L423`](https://github.com/deepmind/mujoco/blob/0e5d062302de3527c6dee766af4efcae78e5f532/sample/simulate.cc#L423) 

Will update the link to the repo.

**Preview:**
* _The sine wave is a data_stream added with `math.sin(x)` inside the main loop_
* _Rest of the lines are joint values_

https://user-images.githubusercontent.com/37873142/170936321-fdd775f7-3dbe-4874-9afe-67eaaebb6c7e.mp4

**Comment by yuvaltassa** [comment]
Excellent!

Very happy you figured this out, and apologies again for the lacking documentation.

**Comment by rohit-kumar-j** [comment]
@yuvaltassa no worries 😅 .Mujoco with python is still new. I'm sure it will take time to fully mature. 

[This branch](https://github.com/rohit-kumar-j/mujoco-python-viewer/tree/extended-example-tutorial) has a rudimentary example in [examples/graphs.py](https://github.com/rohit-kumar-j/mujoco-python-viewer/blob/extended-example-tutorial/examples/graphs.py)

**Reply by yaoxt3** [reply]
Hello, could you please share the entire code? This link has expired. Thanks in advance.

**Reply by rohit-kumar-j** [reply]
@yaoxt3 my fork of mujoco_python_viewer. Although simulate viewer should have a similar functionality now.

https://github.com/rohit-kumar-j/mujoco-python-viewer

**Reply by Apande280** [reply]
What function in the viewer updates the graph? I would like to use one of the joints I have. Can't really figure out which one it is.
